cpp/DEVELOPER_GUIDE.md

```diff
@@ -9,6 +9,61 @@ cuML is thread safe so its functions can be called from multiple host threads if
 
 The implementation of cuML is single threaded.
 
+## Managing state
+### Introduction
+Every ML algo needs to store some state, eg: model and its related hyper-parameters. Thus, this section lays out guidelines for managing state along the API of cuML.
+
+### Inside `libcuml++.so` aka our C++ interface
+Functions exposed via the cuML-C++ layer must be stateless. Meaning, they must accept all the required inputs, parameters and outputs in their argument list only. In other words, the stateful API should always be a wrapper around the stateless methods, NEVER the other way around. That said, internally, these stateless functions are free to create their own temporary classes, as long as they are not exposed on the interface of `libcuml++.so`.
```

Comment:
"[...] in their argument list only" => Maybe add ", which should match the requirements listed below." 

"In other words, the stateful API should always be a wrapper around [...]" => "The stateful API, if one exists, should always be a wrapper around [...]" and/or link to the **scikit-learn-esq stateful API in C++** section below, so a first time reader can see which stateful API we're referring to.

Reply:
Thanks. Done.